### PR TITLE
Charlesmchen/contact share avatars

### PR DESCRIPTION
**Pull request by charlesmchen-signal**
This PR is now ready for review. 

PTAL @michaelkirk

### CODE REVIEW

File: Signal/src/ViewControllers/ContactsPicker.swift [hunk 1]
```diff
@@ -79,7 +79,8 @@ public class ContactsPicker: OWSViewController, UITableViewDelegate, UITableView
         CNContactThumbnailImageDataKey as CNKeyDescriptor,
         CNContactPhoneNumbersKey as CNKeyDescriptor,
         CNContactEmailAddressesKey as CNKeyDescriptor,
-        CNContactPostalAddressesKey as CNKeyDescriptor
+        CNContactPostalAddressesKey as CNKeyDescriptor,
+        CNContactVCardSerialization.descriptorForRequiredKeys()
```
Comment (charlesmchen-signal) [inline Signal/src/ViewControllers/ContactsPicker.swift]: This probably has no effect, but it's nice to be explicit about our required keys.

File: SignalMessaging/ViewModels/ContactShareViewModel.swift [hunk 2]
```diff
@@ -28,9 +28,17 @@ public class ContactShareViewModel: NSObject {
             return avatarImage
         }
 
-        // TODO: What's the best colorSeed value to use?
+        var colorSeed = displayName
```
Comment (charlesmchen-signal) [inline SignalMessaging/ViewModels/ContactShareViewModel.swift]: WDYT?

File: SignalMessaging/attachments/ApproveContactShareViewController.swift [hunk 3]
```diff
@@ -55,6 +55,41 @@ class ContactShareFieldBase<ContactFieldType: OWSContactField>: NSObject, Contac
 
 // MARK: -
 
+// Stub class so that avatars conform to OWSContactField.
+class OWSContactAvatar: NSObject, OWSContactField {
```
Comment (charlesmchen-signal) [inline SignalMessaging/attachments/ApproveContactShareViewController.swift]: Maybe this was a bad idea, but I was trying to maintain the pattern despite some friction.

File: SignalMessaging/attachments/ApproveContactShareViewController.swift [hunk 4]
```diff
@@ -392,11 +436,26 @@ public class ApproveContactShareViewController: OWSViewController, EditContactSh
     // MARK: -
 
     func filteredContactShare() -> ContactShareViewModel {
+        var avatarImage: UIImage?
```
Comment (charlesmchen-signal) [inline SignalMessaging/attachments/ApproveContactShareViewController.swift]: There's definitely some friction around the ViewModel vs. Model in terms of `UIImage` vs. `NSData` and updates.  It'd also be nice to avoid unnecessary image conversions, but we're not doing that - yet.  e.g. if system contact is a jpg, we're reconverting it to jpg when we send the contact share which isn't ideal.

**Comment by charlesmchen-signal** [comment]
This PR is now ready for review. 

There's some open questions about how to safely clean up profile avatars on disk.

PTAL @michaelkirk

File: SignalMessaging/attachments/ApproveContactShareViewController.swift (hunk 4 above)
Comment (michaelkirk-signal) [reply SignalMessaging/attachments/ApproveContactShareViewController.swift]: Right - I wonder if makes sense to move any responsibility that must maintain a reference to the UIImage to the ViewModel, then hopefully we could avoid creating these one-off db connections and creating write transactions on the main thread.

File: SignalServiceKit/src/Messages/Interactions/OWSContact.h [hunk 5]
```diff
@@ -163,13 +164,16 @@ NSString *NSStringForContactAddressType(OWSContactAddressType value);
 
 #pragma mark - System Contact Conversion
 
-+ (nullable OWSContact *)contactForSystemContact:(CNContact *)systemContact;
-+ (nullable CNContact *)systemContactForContact:(OWSContact *)contact;
++ (nullable OWSContact *)contactForSystemContact:(CNContact *)systemContact
+                                     transaction:(YapDatabaseReadWriteTransaction *)transaction;
++ (nullable CNContact *)systemContactForContact:(OWSContact *)contact
+                                    transaction:(YapDatabaseReadTransaction *)transaction;
++ (nullable CNContact *)nameOnlySystemContactForContact:(OWSContact *)contact;
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Messages/Interactions/OWSContact.h]: It doesn't need to be public yet, but it will as soon as we pull `OWSContacts` out to a separate source file (and rename it to something like `OWSContactConversion`), which I'd like to do.

File: SignalMessaging/environment/AppSetup.m [hunk 6]
```diff
@@ -50,6 +50,7 @@ + (void)setupEnvironmentWithCallMessageHandlerBlock:(CallMessageHandlerBlock)cal
         // Register renamed classes.
         [NSKeyedUnarchiver setClass:[OWSUserProfile class] forClassName:[OWSUserProfile collection]];
         [NSKeyedUnarchiver setClass:[OWSDatabaseMigration class] forClassName:[OWSDatabaseMigration collection]];
+        [NSKeyedUnarchiver setClass:[OWSContactShare class] forClassName:@"OWSContact"];
```
Comment (michaelkirk-signal) [inline SignalMessaging/environment/AppSetup.m]: Nice - forgot about this dance.

File: SignalMessaging/profiles/OWSProfileManager.m [hunk 7]
```diff
@@ -765,15 +765,33 @@ - (nullable NSString *)profileNameForRecipientId:(NSString *)recipientId
...
 
     OWSUserProfile *userProfile =
         [OWSUserProfile getOrBuildUserProfileForRecipientId:recipientId dbConnection:self.dbConnection];
 
     if (userProfile.avatarFileName.length > 0) {
-        return [self loadProfileAvatarWithFilename:userProfile.avatarFileName];
+        return [self loadProfileAvatarImageWithFilename:userProfile.avatarFileName];
+    }
+
+    if (userProfile.avatarUrlPath.length > 0) {
```
Comment (michaelkirk-signal) [inline SignalMessaging/profiles/OWSProfileManager.m]: Why this change?

File: SignalServiceKit/src/Messages/DeviceSyncing/OWSIncomingSentMessageTranscript.m [hunk 8]
```diff
@@ -47,7 +47,7 @@ - (instancetype)initWithProto:(OWSSignalServiceProtosSyncMessageSent *)sentProto
 
     _quotedMessage =
         [TSQuotedMessage quotedMessageForDataMessage:_dataMessage thread:_thread relay:relay transaction:transaction];
-    _contact = [OWSContacts contactForDataMessage:_dataMessage relay:relay transaction:transaction];
+    _contactShare = [OWSContactConversion contactShareForDataMessage:_dataMessage relay:relay transaction:transaction];
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Messages/DeviceSyncing/OWSIncomingSentMessageTranscript.m]: much better!

File: SignalServiceKit/src/Messages/Interactions/OWSContactShare.m [hunk 9]
```diff
@@ -262,34 +237,89 @@ - (NSString *)debugDescription
...
+}
+
+- (void)updateDisplayName
+{
+    _displayName = nil;
+
+    [self ensureDisplayName];
+}
+
++ (OWSContactName *)emptyName
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Messages/Interactions/OWSContactShare.m]: Why does this method exist? Why not inline `[OWContactName new]` instead?

File: SignalServiceKit/src/Messages/Interactions/OWSContactShare.m [hunk 10]
```diff
@@ -344,121 +374,80 @@ - (BOOL)ows_isValid
...
-    OWSContact *newContact = [OWSContact new];
+#pragma mark - Phone Numbers and Recipient IDs
 
-    [newContact setNamePrefix:namePrefix
-                    givenName:givenName
-                   middleName:middleName
-                   familyName:familyName
-                   nameSuffix:nameSuffix];
+- (NSArray<NSString *> *)systemContactsWithSignalAccountPhoneNumbers:(id<ContactsManagerProtocol>)contactsManager
+{
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Messages/Interactions/OWSContactShare.m]: Can these be moved to the view model?

File: SignalServiceKit/src/Messages/Interactions/OWSContactShare.m [hunk 11]
```diff
@@ -485,55 +490,67 @@ - (void)saveAvatarImage:(UIImage *)image transaction:(YapDatabaseReadWriteTransa
...
+    _avatarAttachmentId = avatarAttachmentId;
+}
 
-- (NSArray<NSString *> *)systemContactsWithSignalAccountPhoneNumbers:(id<ContactsManagerProtocol>)contactsManager
+@end
+
+#pragma mark -
+
+@implementation OWSContactShareProposed
```
Comment (michaelkirk-signal) [inline SignalServiceKit/src/Messages/Interactions/OWSContactShare.m]: Instead of a subclass of our storage model, why not put the functionality here into the view model?

File: SignalMessaging/attachments/ApproveContactShareViewController.swift (hunk 4 above)
Comment (charlesmchen-signal) [reply SignalMessaging/attachments/ApproveContactShareViewController.swift]: > Right - I wonder if makes sense to move any responsibility that must maintain a reference to the UIImage to the ViewModel, then hopefully we could avoid creating these one-off db connections and creating write transactions on the main thread.

The VM owns the UIImage and I believe we are no longer creating any sync r/w transactions on the main thread.

File: Signal/src/ViewControllers/ConversationView/Cells/OWSContactShareView.m [hunk 12]
```diff
@@ -160,8 +160,9 @@ - (void)createContents
     [contentView autoPinEdgeToSuperviewEdge:ALEdgeTop withInset:self.vMargin];
 
     AvatarImageView *avatarView = [AvatarImageView new];
-    avatarView.image =
-        [self.contactShare getAvatarImageWithDiameter:self.iconSize contactsManager:self.contactsManager];
+    avatarView.image = [ContactShareViewUtils avatarOrDefaultImageForContactShareViewModel:self.contactShare
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/Cells/OWSContactShareView.m]: I've undone this with the base class.

File: Signal/src/ViewControllers/ContactViewController.swift [hunk 13]
```diff
@@ -356,7 +358,7 @@ class ContactViewController: OWSViewController, ContactShareViewHelperDelegate {
             }))
         }
 
-        for email in contactShare.emails {
+        for email in contactShare.emails() {
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ContactViewController.swift]: I've undone this with the base class.

File: Signal/src/ViewControllers/ContactViewController.swift [hunk 14]
```diff
@@ -221,15 +221,17 @@ class ContactViewController: OWSViewController, ContactShareViewHelperDelegate {
...
+                                                                      contactsManager: contactsManager)
         topView.addSubview(avatarView)
         avatarView.autoPin(toTopLayoutGuideOf: self, withInset: 20)
         avatarView.autoHCenterInSuperview()
         avatarView.autoSetDimension(.width, toSize: avatarSize)
         avatarView.autoSetDimension(.height, toSize: avatarSize)
 
         let nameLabel = UILabel()
-        nameLabel.text = contactShare.displayName
+        nameLabel.text = contactShare.displayName()
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ContactViewController.swift]: I've undone this with the base class.

File: Signal/src/ViewControllers/ContactViewController.swift [hunk 15]
```diff
@@ -221,15 +221,17 @@ class ContactViewController: OWSViewController, ContactShareViewHelperDelegate {
 
         let avatarSize: CGFloat = 100
         let avatarView = AvatarImageView()
-        avatarView.image = contactShare.getAvatarImage(diameter: avatarSize, contactsManager: contactsManager)
+        avatarView.image = ContactShareViewUtils.avatarOrDefaultImage(forContactShareViewModel: contactShare,
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ContactViewController.swift]: I've undone this with the base class.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 16]
```diff
@@ -3001,31 +3001,23 @@ - (void)sendMessageAttachment:(SignalAttachment *)attachment
     }
 }
 
-- (void)sendContactShare:(ContactShareViewModel *)contactShare
+- (void)sendContactShare:(OWSContactShare *)contactShare
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: We now convert proposed contact shares to actual contact shares (e.g. write a `TSAttachment`) at the last moment before sending.

We do this:

* ...to avoid cluttering the database with incomplete shares.
* ...to simplify the "cleanup contact share attachments on disk" logic.
* ...to avoid touching the db or creating transactions until we need to.

I've added a separate `sendProposedContactShare` method to make this more clear.

File: Signal/src/ViewControllers/ConversationView/ConversationViewController.m [hunk 17]
```diff
@@ -5002,31 +4994,30 @@ - (void)contactsPicker:(ContactsPicker *)contactsPicker didSelectContact:(Contac
...
+        if (profileAvatarData) {
+            contactShare.avatarData = profileAvatarData;
+            contactShare.isProfileAvatar = YES;
         }
     }
-    contactShareRecord.isProfileAvatar = isProfileAvatar;
 
-    ContactShareViewModel *contactShare =
-        [[ContactShareViewModel alloc] initWithContactShareRecord:contactShareRecord avatarImage:avatarImage];
+    ProposedContactShareViewModel *contactShareVM =
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ConversationView/ConversationViewController.m]: Now `BaseContactShareViewModel` is very close to `ContactShareViewModel` as you had it before.  We use two separate model objects for the reasons above.  They're very similar to each other but differ in how they handle avatars.  We now have `ContactShareViewModel` and `ProposedContactShareViewModel` which for working with these two model types.

Note that the models reside in SSK and the VM reside in SignalMessaging.

File: SignalMessaging/ViewModels/ContactShareViewModel.swift [hunk 18]
```diff
@@ -5,136 +5,227 @@
...
+@objc
+public class ContactShareViewUtils: NSObject {
 
-    public convenience init(contactShareRecord: OWSContact, transaction: YapDatabaseReadTransaction) {
-        if let avatarAttachment = contactShareRecord.avatarAttachment(with: transaction) as? TSAttachmentStream {
-            self.init(contactShareRecord: contactShareRecord, avatarImage: avatarAttachment.image())
-        } else {
-            self.init(contactShareRecord: contactShareRecord, avatarImage: nil)
+    @objc
+    public class func avatarOrDefaultImage(forContactShareViewModel contactShareViewModel: BaseContactShareViewModel,
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply SignalMessaging/ViewModels/ContactShareViewModel.swift]: I've moved it back, now that there's a base class.

File: SignalMessaging/attachments/ApproveContactShareViewController.swift [hunk 19]
```diff
@@ -48,50 +50,88 @@ class ContactShareFieldBase<ContactFieldType: OWSContactField>: NSObject, Contac
...
+        contactShare.avatarData = value.avatarData
+    }
+}
+
+// MARK: -
+
 class ContactSharePhoneNumber: ContactShareFieldBase<OWSContactPhoneNumber> {
 
-    override func applyToContact(contact: ContactShareViewModel) {
+    override func applyToContact(contactShare: OWSContactShareProposed) {
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply SignalMessaging/attachments/ApproveContactShareViewController.swift]: I'm trying to do something analogous to a "builder" pattern where the models are immutable except when they're being constructed.  The VMs now help us enforce this by making the models a private property.  

These `applyToContact ` are all about building a "derived" contact share the reflects the filtering done by the user.

File: SignalMessaging/attachments/ContactFieldView.swift [hunk 20]
```diff
@@ -61,6 +61,31 @@ public class ContactFieldView: UIView {
...
+        stackView.isLayoutMarginsRelativeArrangement = true
+
+        let avatarView = AvatarImageView()
+        avatarView.image = avatarImage
+        let avatarSize = CGFloat(50)
+        avatarView.autoSetDimension(.width, toSize: avatarSize)
+        avatarView.autoSetDimension(.height, toSize: avatarSize)
+        avatarView.setCompressionResistanceHigh()
+        avatarView.setContentHuggingHigh()
+        stackView.addArrangedSubview(avatarView)
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply SignalMessaging/attachments/ContactFieldView.swift]: Because we sometime use `TappableStackView`.  I'm trying to avoid creating a million Tappable classes.

[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply SignalMessaging/attachments/ContactFieldView.swift]: I suppose we could create a UIView+Tappable category.

File: SignalMessaging/attachments/SharingThreadPickerViewController.m [hunk 21]
```diff
@@ -148,25 +148,34 @@ - (void)threadWasSelected:(TSThread *)thread
 
     if (self.attachment.isConvertibleToContactShare) {
         NSData *data = self.attachment.data;
-        OWSContact *_Nullable contact = [OWSContacts contactForVCardData:data];
+        __block OWSContactShareProposed *_Nullable contactShare;
+        [OWSPrimaryStorage.sharedManager.newDatabaseConnection
+            readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply SignalMessaging/attachments/SharingThreadPickerViewController.m]: Changed.

File: SignalMessaging/attachments/SharingThreadPickerViewController.m [hunk 22]
```diff
@@ -282,16 +291,26 @@ - (void)messageApprovalDidCancel:(MessageApprovalViewController *)approvalViewCo
...
 
 - (void)approveContactShare:(ApproveContactShareViewController *)approvalViewController
-     didApproveContactShare:(OWSContact *)contactShare
+     didApproveContactShare:(ProposedContactShareViewModel *)contactShare
 {
     DDLogInfo(@"%@ in %s", self.logTag, __PRETTY_FUNCTION__);
 
+    __block OWSContactShare *convertedContactShare;
+    [OWSPrimaryStorage.sharedManager.newDatabaseConnection
+        readWriteWithBlock:^(YapDatabaseReadWriteTransaction *transaction) {
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply SignalMessaging/attachments/SharingThreadPickerViewController.m]: Changed.

File: SignalMessaging/environment/AppSetup.m (hunk 6 above)
Comment (charlesmchen-signal) [reply SignalMessaging/environment/AppSetup.m]: I tried to add conversions for existing contacts, but it ended up being a nightmare. Once this is merged, you'll probably need to do clean installs on any debug devices.

File: SignalMessaging/profiles/OWSProfileManager.m (hunk 7 above)
Comment (charlesmchen-signal) [reply SignalMessaging/profiles/OWSProfileManager.m]: If we're going to use a profile avatar, we want to access it as _data_, not as an image to avoid an unneccesary and lossy conversion.

File: SignalServiceKit/src/Messages/DeviceSyncing/OWSIncomingSentMessageTranscript.m (hunk 8 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Messages/DeviceSyncing/OWSIncomingSentMessageTranscript.m]: There's more renaming to be done - as you pointed out, we could use contactShareVM in a lot of places now.

File: SignalServiceKit/src/Messages/Interactions/OWSContactShare.m (hunk 9 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Messages/Interactions/OWSContactShare.m]: Removed.

File: SignalServiceKit/src/Messages/Interactions/OWSContactShare.m (hunk 11 above)
Comment (charlesmchen-signal) [reply SignalServiceKit/src/Messages/Interactions/OWSContactShare.m]: There's a few reasons.  

* We want our conversion logic to deal in models, not view models.
* We want to avoid database and file system writes until necessary.

There's a bunch of related and subjective questions here, no doubt.  For one thing, I wanted to avoid a situation where our model's to be used in two completely different ways before/after being sent.

File: Signal/src/ViewControllers/ContactShareViewHelper.swift [hunk 23]
```diff
@@ -170,7 +170,12 @@ public class ContactShareViewHelper: NSObject, CNContactViewControllerDelegate {
             return
         }
 
-        guard let systemContact = OWSContacts.systemContact(for: contactShare.dbRecord) else {
+        var convertedSystemContact: CNContact?
+        OWSPrimaryStorage.shared().newDatabaseConnection().read({ (transaction) in
```
[the remark this answers is not in the document]
Comment (charlesmchen-signal) [reply Signal/src/ViewControllers/ContactShareViewHelper.swift]: I don't see how to avoid that here without making things worse.
